cli/cli_app.go

```diff
@@ -384,7 +384,36 @@ func RunTerragrunt(terragruntOptions *options.TerragruntOptions) error {
 		}
 	}
 
-	return runTerragruntWithConfig(terragruntOptions, terragruntConfig, false)
+	err, output := runTerragruntWithConfig(terragruntOptions, terragruntConfig, false)
+
+	// If running Terraform resulted in an error that matches one of the init errors,
+	// then run the required interim init, and then re-run previous command as-is,
+	// otherwise return error without making any attempt to recover from it or
+	// retry anything.
+	if err != nil && isErrorRequiringInit(output, err, terragruntOptions) {
+		terragruntOptions.Logger.Println("Detected an error that requires Terraform to be re-initialized; running 'terraform init' now...")
+
+		initOptions := terragruntOptions.Clone(terragruntOptions.TerragruntConfigPath)
+		initOptions.Writer = initOptions.ErrWriter
+		initOptions.TerraformCommand = "init"
+		initOptions.TerraformCliArgs = []string{"init"}
+		initOptions.WorkingDir = terragruntOptions.WorkingDir
+		initOptions.AutoInit = false
+
+		// Run the "terraform init" command.
+		if err, _ = runTerragruntWithConfig(initOptions, terragruntConfig, false); err != nil {
+			return err
+		}
```

Comment:
Can we actually use `runTerraformInit` here? That should simplify this routine.

Alternatively, you can use `prepareInitOptions` to get the `terragruntOptions` that works with `init`.

Reply:
I can use `runTerraformInit()`. I gather it's OK to pass `nil` as the last argument (of type `TerraformSource`)?

Reply:
Yup that should be ok.

Reply:
No worries. Would you mind if I clean `runTerraformInit()` a bit? It calls internally `prepareInitOptions()` which takes `TerraformSource` as second argument and then does not do anything with it, and it also returns an error which is always `nil`. As per:

https://github.com/gruntwork-io/terragrunt/blob/94b962bb85b58bee0019c0718c0987bd92179cda/cli/cli_app.go#L748-L759